sql,kvstreamer: use Streamer with multiple column families

yuzefovich opened the pull request:
This commit takes advantage of the recently introduced `WholeRowsOfSize`
argument of the `BatchRequest`s (which allows for SQL rows not being
split across multiple responses when multiple column families are
present) to support multi-column family case in the `Streamer`.

Addresses: #54680.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74921)
<!-- Reviewable:end -->

yuzefovich committed:
sql,kvstreamer: use Streamer with multiple column families

This commit takes advantage of the recently introduced `WholeRowsOfSize`
argument of the `BatchRequest`s (which allows for SQL rows not being
split across multiple responses when multiple column families are
present) to support multi-column family case in the `Streamer`.

Release note: None

rharding6373 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I only have educational questions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74921#-:-MuSjx7H1AY59XBCUo_7:b-pjftpk)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rytaft and @yuzefovich)

___
*[pkg/kv/kvclient/kvstreamer/streamer.go, line 353 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74921#-MuSjoox4WGyG4JJKoFl:-MuSjoox4WGyG4JJKoFm:b-jq308) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51323be78cad0ec4fc4aed954f7516963728114/pkg/kv/kvclient/kvstreamer/streamer.go#L353)):*
> ```Go
> // looked-up keys are unique (e.g. in the case of an index-join).
> //
> // maxKeysPerRow indicates the maximum number of KV pairs that comprise a single
> ```

High level question: Why do we need to now the number of column families in advance?
___
*[pkg/sql/row/kv_batch_streamer.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74921#-MuSj9uABlDd4fx4e4dX:-MuSj9uB3b2gHfBlIR8Q:behh102) ([raw file](https://github.com/cockroachdb/cockroach/blob/ded3357dea64540d3e1eb09b2ffc693a59bd462b/pkg/sql/row/kv_batch_streamer.go#L29)):*
> ```Go
> func CanUseStreamer(ctx context.Context, settings *cluster.Settings) bool {
> 	// TODO(yuzefovich): remove the version gate in 22.2 cycle.
> 	return settings.Version.IsActive(ctx, clusterversion.TargetBytesAvoidExcess) &&
> ```

Why remove the `TargetBytesAvoidExcess` gate?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74921#-:-MuSlk56C1-_DR8LZNS9:b-oduykh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rharding6373 and @rytaft)

___
*[pkg/kv/kvclient/kvstreamer/streamer.go, line 353 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74921#-MuSjoox4WGyG4JJKoFl:-MuSkokX523F6cbhxjAd:b-8tvoj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51323be78cad0ec4fc4aed954f7516963728114/pkg/kv/kvclient/kvstreamer/streamer.go#L353)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

High level question: Why do we need to now the number of column families in advance?
</blockquote></details>

This is so that we could optimize the behavior down in the KV layer. When `WholeRowsOfSize` is greater than one, then the KV will decode the key part of each key-value pair to see whether the current pair belongs to the previous row or to the new row; however, when exactly `WholeRowsOfSize` number of key-values pairs have already been found for the current row, then the row can be "finalized" right away.
___
*[pkg/sql/row/kv_batch_streamer.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74921#-MuSj9uABlDd4fx4e4dX:-MuSlP7TE8Zeasyfoqu2:b3c3y8x) ([raw file](https://github.com/cockroachdb/cockroach/blob/ded3357dea64540d3e1eb09b2ffc693a59bd462b/pkg/sql/row/kv_batch_streamer.go#L29)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Why remove the `TargetBytesAvoidExcess` gate?
</blockquote></details>

`ScanWholeRows` version is larger than `TargetBytesAvoidExcess`, so the presence of the former implies the presence of the latter. Since we haven't had any stable releases yet, there is no point in having some custom code to support single-column-family case in the Streamer, so we just require that the cluster has been upgraded to newer version where any number of column families are supported.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74921)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/kv/kvclient/kvstreamer/streamer.go, line 353 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74921#-MuSjoox4WGyG4JJKoFl:-MuUt_k289ipZ6gkCRhr:b3nehwy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51323be78cad0ec4fc4aed954f7516963728114/pkg/kv/kvclient/kvstreamer/streamer.go#L353)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This is so that we could optimize the behavior down in the KV layer. When `WholeRowsOfSize` is greater than one, then the KV will decode the key part of each key-value pair to see whether the current pair belongs to the previous row or to the new row; however, when exactly `WholeRowsOfSize` number of key-values pairs have already been found for the current row, then the row can be "finalized" right away.
</blockquote></details>

This is true, and important for performance in the case of a single-row scan with lots of MVCC garbage, but not the main reason. The MVCC scanner accumulates results by appending them in a binary encoding to an unstructured byte buffer. To inspect the last SQL row in the result, we need to know where the individual KV pairs start in the byte buffer. Keeping track of the byte offsets of all KV pairs has around 10% overhead plus 8 bytes of memory per pair. Using a fixed-size ring buffer for the last N KV pairs has 2-3% overhead and a constant memory overhead of 8N bytes. But the latter requires us to know the maximum row size, to set the ring buffer size.

https://github.com/cockroachdb/cockroach/blob/6a7678c19d142b8ed6a7ba3c222d3e18aab66c78/pkg/storage/pebble_mvcc_scanner.go#L46-L50

Another case of performance vs. abstraction boundaries. :frowning_face: 


<!-- Sent from Reviewable.io -->
